be2-scala/src/main/scala/ch/epfl/pop/storage/DbActorNew.scala

```diff
@@ -34,24 +34,88 @@ case class DbActorNew(
-    )
+    this.synchronized{
+      val channelData: ChannelData = readChannelData(channel)
+      storage.write(
+        (channel.toString, channelData.addMessage(message.message_id).toJsonString),
+        (s"$channel${Channel.DATA_SEPARATOR}${message.message_id}", message.toJsonString)
+      )
+    }
   }
 
   @throws [DbActorNAckException]
-  private def read(channel: Channel, messageId: Hash): Option[Message] = ???
+  private def read(channel: Channel, messageId: Hash): Option[Message] = {
+    Try(storage.read(s"$channel${Channel.DATA_SEPARATOR}$messageId")) match {
+      case Success(Some(json)) => Some(Message.buildFromJson(json))
+      case Success(None) => None
+      case Failure(ex) => throw DbActorNAckException(ErrorCodes.SERVER_ERROR.id, ex.getMessage)
+    }
+  }
 
   @throws [DbActorNAckException]
-  private def readChannelData(channel: Channel): ChannelData = ???
+  private def readChannelData(channel: Channel): ChannelData = {
+    Try(storage.read(channel.toString)) match {
+      case Success(Some(json)) => ChannelData.buildFromJson(json)
+      case Success(None) => throw DbActorNAckException(ErrorCodes.SERVER_ERROR.id, s"ChannelData for channel $channel not in the database")
+      case Failure(ex) => throw DbActorNAckException(ErrorCodes.SERVER_ERROR.id, ex.getMessage)
+    }
+  }
 
   @throws [DbActorNAckException]
-  private def readLaoData(channel: Channel): LaoData = ???
+  private def readLaoData(channel: Channel): LaoData = {
+    Try(storage.read(generateLaoDataKey(channel))) match {
+      case Success(Some(json)) => LaoData.buildFromJson(json)
+      case Success(None) => LaoData.emptyLaoData
+      case Failure(ex) => throw DbActorNAckException(ErrorCodes.SERVER_ERROR.id, ex.getMessage)
+    }
+  }
 
   @throws [DbActorNAckException]
-  private def catchupChannel(channel: Channel): List[Message] = ???
+  private def writeLaoData(channel: Channel, message: Message): Unit = {
+    this.synchronized{
+      val laoData: LaoData = Try(readLaoData(channel)) match {
+        case Success(data) => data
+        case Failure(_) => LaoData.emptyLaoData
+      }
+      val laoDataKey: String = generateLaoDataKey(channel)
+
+      Try(storage.write(
+        (laoDataKey, laoData.updateWith(message).toJsonString)
+      )).recover(
+        ex => throw DbActorNAckException(ErrorCodes.SERVER_ERROR.id, ex.getMessage)
+      )
+    }
+  }
+
+  @throws [DbActorNAckException]
+  private def catchupChannel(channel: Channel): List[Message] = {
+
+    @scala.annotation.tailrec
+    def buildCatchupList(msgIds: List[Hash], acc: List[Message]): List[Message] = {
+      msgIds match {
+        case Nil => acc
+        case head :: tail =>
+          Try(read(channel, head)).recover(
+            ex => None
+          ) match {
+            case Success(Some(msg)) => buildCatchupList(tail, msg :: acc)
+            case _ => 
+              log.error(s"Problem encountered while catching up message with id $head")
```

Comment:
I would make the message just a little more explicit. If this occurs, it means that the next batch of students f* up somewhere ^^'. We need to make them understand that this should never happen (i.e. we stored a hash in the channel without its associated message). Something like "critical error: channel $channel contains a message that failed to be stored in the db". Maybe even a `log.critical` if it exists with our logger extension

Reply:
EDIT: turns out, there's no log.critical or log.fatal :(

Reply:
EDIT 2 : will be very useful if we add the modifyLao function or delete message for some reason